File: src/controllers/users.controller.ts
import { Request, Response } from 'express';
import { usersService } from '../services/users.service';
import { AuthRequest } from '../types';

function baseUrl(req: Request): string {
  return `${req.protocol}://${req.get('host')}`;
}

export const usersController = {
  async getPublicProfile(req: Request, res: Response) {
    try {
      const data = await usersService.getPublicProfile(req.params.id, baseUrl(req));
      res.json({ data });
    } catch (err: any) {
      res.status(err.status || 500).json({ error: err.code || 'server_error', message: err.message });
    }
  },

  async updateProfile(req: AuthRequest, res: Response) {
    try {
      const data = await usersService.updateProfile(
        req.user!.id,
        { name: req.body.name, phone: req.body.phone },
        baseUrl(req)
      );
      res.json({ data });
    } catch (err: any) {
      res.status(err.status || 500).json({ error: err.code || 'server_error', message: err.message });
    }
  },

  /** PUT /users/me/avatar — multipart/form-data, field name: "avatar" */
  async updateAvatar(req: AuthRequest, res: Response) {
    try {
      if (!req.file) {
        res.status(422).json({ error: 'validation_error', message: 'avatar file is required.' });
        return;
      }
      const data = await usersService.updateAvatar(req.user!.id, req.file, baseUrl(req));
      res.json({ data });
    } catch (err: any) {
      res.status(err.status || 500).json({ error: err.code || 'server_error', message: err.message });
    }
  },

  async changePassword(req: AuthRequest, res: Response) {
    try {
      const { current_password, new_password } = req.body;
      await usersService.changePassword(req.user!.id, current_password, new_password);
      res.json({ message: 'Password changed successfully.' });
    } catch (err: any) {
      res.status(err.status || 500).json({ error: err.code || 'server_error', message: err.message });
    }
  },
};

File: EasyPhones.WebAPI/Controllers/UsersController.cs
using EasyPhones.Application.Dtos.UsersDtos;
using EasyPhones.Application.Services;
using EasyPhones.Application.Wrappers;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace EasyPhones.WebAPI.Controllers;

[ApiController]
[Route("api/users")]
public class UsersController : ControllerBase
{
    private readonly UsersService _users;
    private readonly UserContextService _ctx;

    public UsersController(UsersService users, UserContextService ctx)
    {
        _users = users;
        _ctx = ctx;
    }

    private string BaseUrl() => $"{Request.Scheme}://{Request.Host}";

    /// <summary>Get a user's public profile.</summary>
    [HttpGet("{id}")]
    public async Task<ActionResult<ApiResponse>> GetPublicProfile(string id, CancellationToken ct)
    {
        var profile = await _users.GetPublicProfileAsync(id, BaseUrl(), ct);
        return Ok(ApiResponse.Success(profile));
    }

    /// <summary>Update the authenticated user's name and/or phone.</summary>
    [HttpPut("me")]
    [Authorize]
    public async Task<ActionResult<ApiResponse>> UpdateProfile([FromBody] UpdateProfileDto dto, CancellationToken ct)
    {
        if (!ModelState.IsValid) return ValidationProblem(ModelState);
        var profile = await _users.UpdateProfileAsync(_ctx.UserId, dto, BaseUrl(), ct);
        return Ok(ApiResponse.Success(profile));
    }

    /// <summary>Upload a new avatar (multipart/form-data, field "avatar").</summary>
    [HttpPut("me/avatar")]
    [Authorize]
    [RequestSizeLimit(5 * 1024 * 1024)]
    public async Task<ActionResult<ApiResponse>> UpdateAvatar(CancellationToken ct)
    {
        var file = Request.Form.Files.GetFile("avatar");
        if (file is null)
            return UnprocessableEntity(ApiResponse.Fail("validation_error", "avatar file is required."));

        var allowed = new[] { "image/jpeg", "image/jpg", "image/png", "image/webp", "image/gif" };
        if (!allowed.Contains(file.ContentType))
            return UnprocessableEntity(ApiResponse.Fail("validation_error",
                "Only image files are allowed (jpeg, png, webp, gif)."));

        using var ms = new MemoryStream();
        await file.CopyToAsync(ms);

        var profile = await _users.UpdateAvatarAsync(
            _ctx.UserId, file.FileName, file.ContentType, file.Length, ms.ToArray(), BaseUrl(), ct);

        return Ok(ApiResponse.Success(profile));
    }

    /// <summary>Change the authenticated user's password.</summary>
    [HttpPut("me/password")]
    [Authorize]
    public async Task<ActionResult<ApiResponse>> ChangePassword([FromBody] ChangePasswordDto dto, CancellationToken ct)
    {
        if (!ModelState.IsValid) return ValidationProblem(ModelState);
        await _users.ChangePasswordAsync(_ctx.UserId, dto, ct);
        return Ok(ApiResponse.Success(message: "Password changed successfully."));
    }
}
